### ISSUE TITLE
object_detection evaluator graph_hook_fn() might insert model variables to be restored

### ISSUE THREAD

**Issue by DominikAuras** [tensorflow/models#3611]
The graph hook function, executed in `trainer.py`
https://github.com/tensorflow/models/blob/c31b3c2c883aadffc37bc33643fa727185f59ce2/research/object_detection/trainer.py#L317-L319
and `evaluator.py`
https://github.com/tensorflow/models/blob/c31b3c2c883aadffc37bc33643fa727185f59ce2/research/object_detection/evaluator.py#L216
might actually insert model variables that should be restored from checkpoints. 

Example: `tf.contrib.quantize` inserts `min/max` variables that represent the quantization range.
https://github.com/tensorflow/tensorflow/blob/336ca16154dc062f2cbab56502395f14f4d01e49/tensorflow/contrib/quantize/python/quant_ops.py#L98-L109

The package `tf.contrib.quantize` rewrites the graph, inserting fake quantization operations, and one way to use it is to pack it into the graph hook function.

Due to this code snippet here
https://github.com/tensorflow/models/blob/c31b3c2c883aadffc37bc33643fa727185f59ce2/research/object_detection/evaluator.py#L212-L216
the graph hook in the evaluator is executed *after* determining which variables to restore, hence the `graph_hook_fn` inserted model variables, such as `min/max` from `tf.contrib.quantize`, are never restored. This was difficult to find, as there is no warning about variables in the checkpoint that are not restored, and by default, during evaluation no summaries of model variables are evaluated, i.e., they are not inspectable in Tensorboard. I noticed this after modifying the evaluator to show me the summaries of variables from the evaluation run in Tensorboard.

Proposed solution
```python
 if graph_hook_fn: graph_hook_fn() 

 variables_to_restore = tf.global_variables() 
 global_step = tf.train.get_or_create_global_step() 
 variables_to_restore.append(global_step) 
```

**Comment by coutner** [comment]
@DominikAuras how did you pack tf.contrib.quantize into the graph hook function?I just replace graph_hook_fn() with tf.contrib.quantize.create_training_graph().But when I inspect the checkpoint after quantization,I find a duplicate act_quant tensor name.

> 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/act_quant/FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/act_quant/min/biased'
'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/act_quant/min'

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by rinley** [comment]
@DominikAuras Thanks. This is very helpful. It is a bug in my opinion. I tried your proposed solution and it worked.

Also, if you are using docker (Latest tensorflow-py3 docker as of today, 06/05/2018) and trying to use the fake quantization for training SSD+MobileNetV2, you would probably encounter the problem of unmatched dimension of batchnorm op. I solved it by updating the tensorflow/contrib/quantize/python/fold_batch_norms.py to this commit:283e8fe7e191f8e0e2ca6ea62b8b4553c30a6286. https://github.com/tensorflow/tensorflow/commit/283e8fe7e191f8e0e2ca6ea62b8b4553c30a6286#diff-17d3ec587749ec8343e812c1b77e7309
